File: python/src/smooth/adam_general/core/auto_om.py
"""Automatic Occurrence Model Selection (AutoOM): port of R's ``auto.om()``.

Tries each occurrence type from ``occurrence`` and returns the model with the
lowest information criterion.
"""

from __future__ import annotations

import time
import warnings
from typing import Any, Dict, List, Literal, Optional, Union

from numpy.typing import NDArray

from smooth.adam_general.core.om import OM
from smooth.adam_general.core.omg import OMG

_VALID_OCCURRENCE = (
    "fixed",
    "odds-ratio",
    "inverse-odds-ratio",
    "direct",
    "general",
)

_IC_ATTR = {"AIC": "aic", "AICc": "aicc", "BIC": "bic", "BICc": "bicc"}


def _get_ic(model: Union[OM, OMG], ic_name: str) -> float:
    return float(getattr(model, _IC_ATTR[ic_name]))


class AutoOM:
    """Automatic occurrence model selection — port of R's ``auto.om()``.

    Fits an :class:`OM` (or :class:`OMG`) for each entry in ``occurrence``
    and returns the best-fitting model (lowest IC) directly.  The returned
    object is a plain :class:`OM` or :class:`OMG` with one extra attribute
    ``time_elapsed_`` (total selection time in seconds), matching R's
    ``$timeElapsed`` on the returned ``om`` object.

    Parameters
    ----------
    model :
        ETS spec passed to every candidate, including both sub-models of the
        general (OMG) candidate.
    lags :
        Seasonal lags shared across all candidates.
    orders :
        ARIMA orders forwarded to every candidate.
    occurrence :
        Sequence of occurrence types to try.
    h, holdout :
        Forecast horizon and holdout flag forwarded to every candidate.
    ic :
        Information criterion used for selection.
    """

    def __init__(
        self,
        model: str = "ZXZ",
        lags: Optional[List[int]] = None,
        orders: Optional[Dict[str, Any]] = None,
        occurrence: Union[List[str], str] = list(_VALID_OCCURRENCE),
        h: int = 0,
        holdout: bool = False,
        persistence: Optional[Dict[str, float]] = None,
        phi: Optional[float] = None,
        initial: Union[str, Dict[str, Any]] = "backcasting",
        constant: bool = False,
        arma: Optional[Dict[str, Any]] = None,
        regressors: Literal["use", "select", "adapt"] = "use",
        ic: Literal["AIC", "AICc", "BIC", "BICc"] = "AICc",
        bounds: Literal["usual", "admissible", "none"] = "usual",
        verbose: int = 0,
        nlopt_kargs: Optional[Dict[str, Any]] = None,
        ets: Literal["conventional", "adam"] = "conventional",
    ) -> None:
        if isinstance(occurrence, str):
            occurrence = [occurrence]
        unknown = set(occurrence) - set(_VALID_OCCURRENCE)
        if unknown:
            raise ValueError(f"Unknown occurrence type(s): {unknown!r}")
        if ic not in _IC_ATTR:
            raise ValueError(f"ic must be one of {list(_IC_ATTR)}; got {ic!r}")
        self.model = model
        self.lags = lags
        self.orders = orders
        self.occurrence = list(occurrence)
        self.h = h
        self.holdout = holdout
        self.persistence = persistence
        self.phi = phi
        self.initial = initial
        self.constant = constant
        self.arma = arma
        self.regressors = regressors
        self.ic = ic
        self.bounds = bounds
        self.verbose = verbose
        self.nlopt_kargs = nlopt_kargs
        if ets not in ("conventional", "adam"):
            raise ValueError(f"Invalid ets: {ets!r}. Must be 'conventional' or 'adam'.")
        self.ets = ets

    # ------------------------------------------------------------------
    # Private helpers
    # ------------------------------------------------------------------

    def _common_kwargs(self) -> Dict[str, Any]:
        return dict(
            lags=self.lags,
            h=self.h,
            holdout=self.holdout,
            initial=self.initial,
            ic=self.ic,
            bounds=self.bounds,
            verbose=self.verbose,
            nlopt_kargs=self.nlopt_kargs,
        )

    def _build_candidate(self, occ: str) -> Union[OM, OMG]:
        common = self._common_kwargs()
        if occ == "general":
            return OMG(
                model_a=self.model,
                model_b=self.model,
                orders_a=self.orders,
                constant_a=self.constant,
                persistence_a=self.persistence,
                phi_a=self.phi,
                arma_a=self.arma,
                regressors_a=self.regressors,
                ets=self.ets,
                **common,
            )
        return OM(
            model=self.model,
            occurrence=occ,
            orders=self.orders,
            constant=self.constant,
            persistence=self.persistence,
            phi=self.phi,
            arma=self.arma,
            regressors=self.regressors,
            ets=self.ets,
            **common,
        )

    # ------------------------------------------------------------------
    # fit
    # ------------------------------------------------------------------

    def fit(self, y: NDArray, X: Optional[NDArray] = None) -> Union[OM, OMG]:
        """Fit all candidates and return the best model by IC.

        Parameters
        ----------
        y :
            Observation vector (binarised internally: any non-zero → 1).
        X :
            Optional regressor matrix forwarded to every candidate.

        Returns
        -------
        OM or OMG
            The best-fitting occurrence model, with ``time_elapsed_`` set to
            the total selection time in seconds.
        """
        start = time.time()
        best_ic = float("inf")
        best_model: Optional[Union[OM, OMG]] = None

        for occ in self.occurrence:
            try:
                m = self._build_candidate(occ)
                m.fit(y, X)
                ic_val = _get_ic(m, self.ic)
                if ic_val < best_ic:
                    best_ic = ic_val
                    best_model = m
            except Exception as exc:
                warnings.warn(
                    f"AutoOM: occurrence='{occ}' failed with {type(exc).__name__}: "
                    f"{exc} — skipping.",
                    stacklevel=2,
                )

        if best_model is None:
            raise RuntimeError("AutoOM: all occurrence types failed to fit.")

        best_model.time_elapsed_ = time.time() - start
        return best_model

File: R/auto-om.R
#' Automatic Occurrence Model Selection
#'
#' Fits \link{om} for each supplied \code{occurrence} type and returns the
#' model with the lowest information criterion.
#'
#' @param data Numeric vector, time series, or data frame. Non-binary input is
#'   automatically binarised: any non-zero value becomes 1.
#' @param model Three-letter ETS specification (wildcards \code{"Z"}/\code{"X"}/\code{"Y"} supported).
#' @param lags Vector of seasonal lags. Defaults to \code{frequency(data)}.
#' @param orders ARIMA orders: \code{list(ar, i, ma, select=TRUE/FALSE)}.
#'   When \code{select=TRUE}, ARIMA orders are selected automatically for each
#'   occurrence type.
#' @param formula Optional formula for external regressors.
#' @param regressors How to handle regressors: \code{"use"}, \code{"select"}, or \code{"adapt"}.
#' @param occurrence Character vector of occurrence link types to try:
#'   \code{"fixed"}, \code{"odds-ratio"}, \code{"inverse-odds-ratio"}, \code{"direct"}.
#' @param h Forecast horizon.
#' @param holdout If \code{TRUE}, a holdout sample of size \code{h} is withheld.
#' @param persistence Optional persistence (smoothing) parameter vector.
#' @param phi Optional damping parameter.
#' @param initial Initialisation method: \code{"backcasting"}, \code{"optimal"},
#'   \code{"two-stage"}, or \code{"complete"}.
#' @param arma Optional fixed ARMA parameters.
#' @param ic Information criterion used for selection.
#' @param bounds Parameter bounds type.
#' @param silent If \code{TRUE}, suppresses progress messages and plots.
#' @param parallel If \code{TRUE} (or a core count), fit occurrence types in
#'   parallel.  Requires \code{foreach} and \code{doMC} / \code{doParallel}.
#' @param ets ETS flavour passed to \link{om}: \code{"conventional"} or \code{"adam"}.
#' @param ... Additional arguments forwarded to \link{om}.
#'
#' @return The best \code{om} object (lowest IC), with \code{$call} set to
#'   the \code{auto.om()} call and \code{$timeElapsed} recording wall time.
#'
#' @seealso \link{om}, \link{auto.adam}
#'
#' @examples
#' set.seed(42)
#' y <- rbinom(120, 1, 0.6)
#' \donttest{m <- auto.om(y, occurrence = c("fixed", "odds-ratio"))}
#'
#' @export
auto.om <- function(data,
                    model      = "ZXZ",
                    lags       = c(frequency(data)),
                    orders     = list(ar=c(3,3), i=c(2,1), ma=c(3,3), select=TRUE),
                    formula    = NULL,
                    regressors = c("use","select","adapt"),
                    occurrence = c("fixed","odds-ratio","inverse-odds-ratio","direct","general"),
                    h = 0, holdout = FALSE,
                    persistence = NULL, phi = NULL,
                    initial = c("backcasting","optimal","two-stage","complete"),
                    arma = NULL,
                    ic      = c("AICc","AIC","BIC","BICc"),
                    bounds  = c("usual","admissible","none"),
                    silent  = TRUE,
                    parallel = FALSE,
                    ets     = c("conventional","adam"),
                    ...){
    startTime <- Sys.time();
    cl <- match.call();

    #### IC / matching ####
    ic      <- match.arg(ic, c("AICc","AIC","BIC","BICc"));
    IC      <- switch(ic, "AIC"=AIC, "AICc"=AICc, "BIC"=BIC, "BICc"=BICc);
    initial <- match.arg(initial);
    ets     <- match.arg(ets);
    regressors <- match.arg(regressors);

    occurrence <- match.arg(occurrence,
                            c("fixed","odds-ratio","inverse-odds-ratio","direct","general"),
                            several.ok=TRUE);

    if(any(unlist(strsplit(model,""))=="C")){
        modelDo <- "combine";
    }
    else{
        modelDo <- "select";
    }

    #### Data prep ####
    if(is.adam.sim(data) || is.smooth.sim(data)){
        data <- data$data;
    }
    else if(inherits(data,"Mdata")){
        h <- data$h;
        holdout <- TRUE;
        lags <- frequency(data$x);
        data <- ts(c(data$x, data$xx), start=start(data$x), frequency=frequency(data$x));
    }

    if(is.null(dim(data))){
        obsInSample <- length(data) - holdout*h;
    }
    else{
        obsInSample <- nrow(data) - holdout*h;
        if(is.null(formula)){
            responseName <- colnames(data)[1];
        }
        else{
            responseName <- all.vars(formula)[1];
        }
    }

    if(!is.null(dim(data))){
        yInSample <- if(is.null(formula)) data[1:obsInSample, 1] else
            data[1:obsInSample, all.vars(formula)[1]];
    }
    else{
        yInSample  <- data[1:obsInSample];
        responseName <- paste0(deparse(substitute(data)), collapse="");
        responseName <- make.names(responseName, unique=TRUE);
    }

    #### Model flags ####
    etsModel  <- all(model != "NNN");
    xregModel <- (!is.null(dim(data)) && ncol(data) > 1);

    #### ARIMA parameter processing ####
    if(is.list(orders)){
        arimaModel       <- any(c(orders$ar, orders$i, orders$ma) > 0);
    }
    else{
        arimaModel <- any(orders > 0);
    }

    if(arimaModel){
        arimaModelSelect <- FALSE;
        if(is.list(orders)){
            arimaModelSelect <- isTRUE(orders$select);
            arMax <- orders$ar;
            iMax  <- orders$i;
            maMax <- orders$ma;
        }
        else{
            arMax <- orders[1];
            iMax  <- orders[2];
            maMax <- orders[3];
        }

        if(any(c(arMax, iMax, maMax) < 0)){
            stop("Funny guy! How am I gonna construct a model with negative order?", call.=FALSE);
        }

        if(sum(lags==1) == 0){
            lags <- c(1, lags);
        }
        if(any(lags==0)){
            arMax <- arMax[lags!=0];
            iMax  <- iMax[lags!=0];
            maMax <- maMax[lags!=0];
            lags  <- lags[lags!=0];
        }

        maxorder <- max(length(arMax), length(iMax), length(maMax), length(lags));
        if(length(arMax) != maxorder){ arMax <- c(arMax, rep(0, maxorder-length(arMax))); }
        if(length(iMax)  != maxorder){ iMax  <- c(iMax,  rep(0, maxorder-length(iMax)));  }
        if(length(maMax) != maxorder){ maMax <- c(maMax, rep(0, maxorder-length(maMax))); }

        if(any((arMax + iMax + maMax) == 0)){
            orders2leave <- (arMax + iMax + maMax) != 0;
            if(all(!orders2leave)){ orders2leave <- lags == min(lags); }
            arMax <- arMax[orders2leave];
            iMax  <- iMax[orders2leave];
            maMax <- maMax[orders2leave];
            lags  <- lags[orders2leave];
        }

        if(length(unique(lags)) != length(lags)){
            lagsNew  <- unique(lags);
            arMaxNew <- iMaxNew <- maMaxNew <- lagsNew;
            for(i in seq_along(lagsNew)){
                arMaxNew[i] <- max(arMax[which(lags==lagsNew[i])], na.rm=TRUE);
                iMaxNew[i]  <- max(iMax[which(lags==lagsNew[i])],  na.rm=TRUE);
                maMaxNew[i] <- max(maMax[which(lags==lagsNew[i])], na.rm=TRUE);
            }
            arMax <- arMaxNew;
            iMax  <- iMaxNew;
            maMax <- maMaxNew;
            lags  <- lagsNew;
        }

        arMax <- arMax[order(lags)];
        iMax  <- iMax[order(lags)];
        maMax <- maMax[order(lags)];
        lags  <- sort(lags);
    }
    else{
        arMax <- iMax <- maMax <- NULL;
        arimaModelSelect <- FALSE;
        orders <- c(0,0,0);
    }

    nModels <- length(occurrence);

    #### Parallel setup ####
    setupResult <- adam_setupParallel(parallel, nModels);
    parallel    <- setupResult$parallel;
    cluster     <- setupResult$cluster;

    if(!silent){
        if(!parallel){
            cat("Evaluating occurrence models... ");
        }
        else{
            cat("Working... ");
        }
    }

    #### omReturner: fits om for each occurrence type ####
    omReturner <- function(data, model, lags, orders,
                           occurrence, h, holdout,
                           persistence, phi, initial, arma,
                           ic, bounds, regressors, parallel,
                           arimaModelSelect, arMax, iMax, maMax, ...){
        ordersToUse <- if(arimaModelSelect) list(ar=0, i=0, ma=0, select=FALSE) else orders;

        if(!parallel){
            selectedModels <- vector("list", length(occurrence));
            for(i in seq_along(occurrence)){
                if(!silent){ cat(occurrence[i], "\b, "); }
                if(etsModel || xregModel || (arimaModel && !arimaModelSelect) ||
                   occurrence[i] == "general"){
                    selectedModels[[i]] <- om(data=data, model=model, lags=lags,
                                              orders=ordersToUse,
                                              occurrence=occurrence[i], formula=formula,
                                              h=h, holdout=holdout,
                                              persistence=persistence, phi=phi,
                                              initial=initial, arma=arma,
                                              ic=ic, bounds=bounds,
                                              regressors=regressors,
                                              silent=TRUE, ets=ets, ...);
                }
                if(arimaModelSelect && occurrence[i] != "general"){
                    selectedModels[[i]] <- adam_arimaSelector(
                        data=data, model=model,
                        lags=lags, arMax=arMax, iMax=iMax, maMax=maMax,
                        h=h, holdout=holdout,
                        persistence=persistence, phi=phi, initial=initial,
                        ic=ic, bounds=bounds, silent=silent, regressors=regressors,
                        testModelETS=selectedModels[[i]],
                        fitter=om, fitter_args=list(occurrence=occurrence[i]),
                        IC=IC, formula=formula, ets=ets,
                        responseName=responseName, obsInSample=obsInSample, ...);
                }
            }
        }
        else{
            selectedModels <- foreach::`%dopar%`(
                foreach::foreach(i=seq_along(occurrence)),
                {
                    if(etsModel || xregModel || occurrence[i] == "general"){
                        testModel <- om(data=data, model=model, lags=lags,
                                        orders=ordersToUse,
                                        occurrence=occurrence[i], formula=formula,
                                        h=h, holdout=holdout,
                                        persistence=persistence, phi=phi,
                                        initial=initial, arma=arma,
                                        ic=ic, bounds=bounds,
                                        regressors=regressors,
                                        silent=TRUE, ets=ets, ...);
                    }
                    else{
                        testModel <- NULL;
                    }
                    if(arimaModelSelect && occurrence[i] != "general"){
                        testModel <- adam_arimaSelector(
                            data=data, model=model,
                            lags=lags, arMax=arMax, iMax=iMax, maMax=maMax,
                            h=h, holdout=holdout,
                            persistence=persistence, phi=phi, initial=initial,
                            ic=ic, bounds=bounds, silent=TRUE, regressors=regressors,
                            testModelETS=testModel,
                            fitter=om, fitter_args=list(occurrence=occurrence[i]),
                            IC=IC, formula=formula, ets=ets,
                            responseName=responseName, obsInSample=obsInSample, ...);
                    }
                    return(testModel);
                });
        }
        return(selectedModels);
    }

    if(arimaModelSelect && !is.null(arma)){
        warning("ARIMA order selection cannot be done with provided arma parameters. Dropping them.",
                call.=FALSE);
        arma <- NULL;
    }

    #### Run models ####
    if(!arimaModelSelect){
        selectedModels <- omReturner(data, model, lags, orders,
                                     occurrence, h, holdout,
                                     persistence, phi, initial, arma,
                                     ic, bounds, regressors, parallel,
                                     arimaModelSelect, arMax, iMax, maMax, ...);
    }
    else if(etsModel || xregModel){
        selectedModels <- omReturner(data, model, lags, orders,
                                     occurrence, h, holdout,
                                     persistence, phi, initial, arma,
                                     ic, bounds, regressors, parallel,
                                     arimaModelSelect, arMax, iMax, maMax, ...);
    }
    else{
        if(!parallel){
            selectedModels <- vector("list", length(occurrence));
            for(i in seq_along(occurrence)){
                if(!silent){ cat(occurrence[i], "\b: "); }
                if(occurrence[i] != "general"){
                    selectedModels[[i]] <- adam_arimaSelector(
                        data=data, model=model,
                        lags=lags, arMax=arMax, iMax=iMax, maMax=maMax,
                        h=h, holdout=holdout,
                        persistence=persistence, phi=phi, initial=initial,
                        ic=ic, bounds=bounds, silent=silent, regressors=regressors,
                        testModelETS=NULL,
                        fitter=om, fitter_args=list(occurrence=occurrence[i]),
                        IC=IC, formula=formula, ets=ets,
                        responseName=responseName, obsInSample=obsInSample, ...);
                } else {
                    selectedModels[[i]] <- om(data=data, model=model, lags=lags,
                                              orders=c(0,0,0), occurrence="general",
                                              formula=formula, h=h, holdout=holdout,
                                              persistence=persistence, phi=phi,
                                              initial=initial, arma=arma, ic=ic, bounds=bounds,
                                              regressors=regressors, silent=TRUE, ets=ets, ...);
                }
            }
        }
        else{
            selectedModels <- foreach::`%dopar%`(
                foreach::foreach(i=seq_along(occurrence)),
                {
                    if(occurrence[i] != "general"){
                        testModel <- adam_arimaSelector(
                            data=data, model=model,
                            lags=lags, arMax=arMax, iMax=iMax, maMax=maMax,
                            h=h, holdout=holdout,
                            persistence=persistence, phi=phi, initial=initial,
                            ic=ic, bounds=bounds, silent=TRUE, regressors=regressors,
                            testModelETS=NULL,
                            fitter=om, fitter_args=list(occurrence=occurrence[i]),
                            IC=IC, formula=formula, ets=ets,
                            responseName=responseName, obsInSample=obsInSample, ...);
                    } else {
                        testModel <- om(data=data, model=model, lags=lags,
                                        orders=c(0,0,0), occurrence="general",
                                        formula=formula, h=h, holdout=holdout,
                                        persistence=persistence, phi=phi,
                                        initial=initial, arma=arma, ic=ic, bounds=bounds,
                                        regressors=regressors, silent=TRUE, ets=ets, ...);
                    }
                    return(testModel);
                });
        }
    }

    #### Select best ####
    safeIC <- function(m){
        if(!is.null(m$ICw) && !is.null(m$ICs) && is.numeric(m$ICs) && is.numeric(m$ICw)){
            return(as.numeric(m$ICs[is.finite(m$ICs)] %*% m$ICw[is.finite(m$ICs)]));
        }
        tryCatch(IC(m), warning=function(w) Inf, error=function(e) Inf);
    }
    ICValues <- vapply(selectedModels, safeIC, numeric(1));

    best <- which.min(ICValues);
    selectedModels[[best]]$timeElapsed <- Sys.time() - startTime;
    selectedModels[[best]]$call <- cl;

    if(!silent){
        cat("Done!\n");
        plot(selectedModels[[best]], 7);
    }

    adam_teardownParallel(cluster);

    return(selectedModels[[best]]);
}
